t/discovery/stream/dns.t

```diff
@@ -0,0 +1,431 @@
+--- error_log
+failed to query the DNS server
+
+
+
+=== TEST 7: SRV
+--- apisix_yaml
+upstreams:
+    - service_name: "srv.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- grep_error_log eval
+qr/upstream nodes: \{[^}]+\}/
+--- grep_error_log_out eval
+qr/upstream nodes: \{("127.0.0.1:1980":60,"127.0.0.2:1980":20|"127.0.0.2:1980":20,"127.0.0.1:1980":60)\}/
+--- stream_response_like
+hello world
+
+
+
+=== TEST 8: SRV (RFC 2782 style)
+--- apisix_yaml
+upstreams:
+    - service_name: "_sip._tcp.srv.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- grep_error_log eval
+qr/upstream nodes: \{[^}]+\}/
+--- grep_error_log_out eval
+qr/upstream nodes: \{("127.0.0.1:1980":60,"127.0.0.2:1980":20|"127.0.0.2:1980":20,"127.0.0.1:1980":60)\}/
+--- stream_response_like
+hello world
+
+
+
+=== TEST 9: SRV (different port)
+--- apisix_yaml
+upstreams:
+    - service_name: "port.srv.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- grep_error_log eval
+qr/upstream nodes: \{[^}]+\}/
+--- grep_error_log_out eval
+qr/upstream nodes: \{("127.0.0.1:1980":60,"127.0.0.2:1981":20|"127.0.0.2:1981":20,"127.0.0.1:1980":60)\}/
+--- stream_response_like
+hello world
+
+
+
+=== TEST 10: SRV (zero weight)
+--- apisix_yaml
+upstreams:
+    - service_name: "zero-weight.srv.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- grep_error_log eval
+qr/upstream nodes: \{[^}]+\}/
+--- grep_error_log_out eval
+qr/upstream nodes: \{("127.0.0.1:1980":60,"127.0.0.2:1980":1|"127.0.0.2:1980":1,"127.0.0.1:1980":60)\}/
+--- stream_response_like
+hello world
+
+
+
+=== TEST 11: SRV (split weight)
+--- apisix_yaml
+upstreams:
+    - service_name: "split-weight.srv.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- grep_error_log eval
+qr/upstream nodes: \{[^}]+\}/
+--- grep_error_log_out eval
+qr/upstream nodes: \{(,?"127.0.0.(1:1980":200|3:1980":1|4:1980":1)){3}\}/
+--- stream_response_like
+hello world
+
+
+
+=== TEST 12: SRV (priority)
+--- apisix_yaml
+upstreams:
+    - service_name: "priority.srv.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- stream_response_like
+hello world
+--- error_log
+connect() failed
+--- grep_error_log eval
+qr/proxy request to \S+/
+--- grep_error_log_out
+proxy request to 127.0.0.1:1979
+proxy request to 127.0.0.2:1980
+
+
+
+=== TEST 13: prefer SRV than A
+--- apisix_yaml
+upstreams:
+    - service_name: "srv-a.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- error_log
+proxy request to 127.0.0.1:1980
+--- stream_response_like
+hello world
+
+
+
+=== TEST 14: SRV (port is 0)
+--- apisix_yaml
+upstreams:
+    - service_name: "zero.srv.test.local"
+      discovery_type: dns
+      type: roundrobin
+      id: 1
+--- error_log
+connect() failed
+--- grep_error_log eval
+qr/proxy request to \S+/
+--- grep_error_log_out
+proxy request to 127.0.0.1:nil
```

Comment:
For L7, we will fill node info in `apisix/upstream.lua` https://github.com/apache/apisix/blob/030019b5297268b2171df5aa661d84c3c1ab6b92/apisix/upstream.lua#L178

For L4, if node port is 0, it will not be filled node info, so the error log is like `proxy request to 127.0.0.1:nil` which in the `apisix/balancer.lua` https://github.com/apache/apisix/blob/030019b5297268b2171df5aa661d84c3c1ab6b92/apisix/balancer.lua#L384

if we need to reject zero port in the stream, i wonder we should reject in where, can you give me some hint? thank you.

Reply:
i ignore zero port node when is in the stream subsystem.